kafka-connector/src/main/java/com/google/pubsub/kafka/sink/CloudPubSubSinkTask.java

```diff
@@ -118,12 +128,31 @@ public void put(Collection<SinkRecord> sinkRecords) {
       ByteString value = handleValue(record.valueSchema(), record.value(), attributes);
       // Get the total number of bytes in this message.
       int messageSize = value.size(); // Assumes the topic name is in ASCII.
-      if (record.key() != null) {
-        attributes.put(ConnectorUtils.CPS_MESSAGE_KEY_ATTRIBUTE, record.key().toString());
-      }
       for (String key : attributes.keySet()) {
         messageSize+= key.getBytes().length + attributes.get(key).getBytes().length;
       }
+      if (record.key() != null) {
+        String key = record.key().toString();
+        attributes.put(ConnectorUtils.CPS_MESSAGE_KEY_ATTRIBUTE, key);
+        messageSize += CPS_MESSAGE_KEY_ATTRIBUTE_SIZE + key.getBytes().length;
+      }
+      if (includeMetadata) {
+        String topic = record.topic().toString();
+        attributes.put(ConnectorUtils.KAFKA_TOPIC_ATTRIBUTE, topic);
+        messageSize += KAFKA_TOPIC_ATTRIBUTE_SIZE + topic.getBytes().length;
+
+        String partition = record.kafkaPartition().toString();
+        attributes.put(ConnectorUtils.KAFKA_PARTITION_ATTRIBUTE, partition);
+        messageSize += KAFKA_PARTITION_ATTRIBUTE_SIZE + partition.getBytes().length;
+
+        String offset = Long.toString(record.kafkaOffset());
+        attributes.put(ConnectorUtils.KAFKA_OFFSET_ATTRIBUTE, offset);
+        messageSize += KAFKA_OFFSET_ATTRIBUTE_SIZE + offset.getBytes().length;
+
+        String timestamp = record.timestamp().toString();
+        attributes.put(ConnectorUtils.KAFKA_TIMESTAMP_ATTRIBUTE, timestamp);
+        messageSize += KAFKA_TIMESTAMP_ATTRIBUTE_SIZE + timestamp.getBytes().length;
```

Comment:
is it not easier to just calculate messageSize by calling getSerializedSize on the proto at the end? This seems brittler since you have to constantly keep track of it

Reply:
Done!